src/KlaviyoBase.php

```diff
@@ -0,0 +1,292 @@
+    public function __construct( $public_key, $private_key )
+    {
+        $this->public_key = $public_key;
+        $this->private_key = $private_key;
+        $this->client = new Client(['base_uri' => self::HOST]);
+    }
+
+    public function getPublickey()
+    {
+        return $this->public_key;
+    }
+
+    public function getPrivateKey()
+    {
+        return $this->private_key;
+    }
+
+    /**
+     * Make public API request
+     */
+    protected function publicRequest( $path, $options )
+    {
+        // Public requests are always GET
+        return $this->request( self::HTTP_GET, $path, $options, true );
+    }
+
+    /**
+     * Make private v1 API request
+     */
+    protected function v1Request( $path, $options = [], $method = self::HTTP_GET )
+    {
+        $path = self::V1 . $this->trimPath( $path );
+
+        return $this->request( $method, $path, $options, false, true );
+    }
+
+    /**
+     * Make private v2 API request
+     */
+    protected function v2Request( $path, $options = [], $method = self::HTTP_GET )
+    {
+        $path = self::V2 . $this->trimPath( $path );
+
+        return $this->request( $method, $path, $options, false, false, true );
+    }
+
+    /**
+     * Make API request using HTTP client
+     */
+    private function request( $method, $path, $options, $isPublic = false, $isV1 = false, $isV2 = false )
+    {
+        $this->prepareAuthentication( $options, $isPublic, $isV1, $isV2 );
+
+        $response = $this->client->request( $method, $path, $options );
+
+        return $this->handleResponse( $response, $isPublic );
+    }
+
+    /**
+     * Handle response from API call
+     */
+    private function handleResponse( ResponseInterface $response, $isPublic )
+    {
+        $statusCode = $response->getStatusCode();
+
+        if ( $statusCode == 403 ) {
+            throw new KlaviyoAuthenticationException(self::ERROR_INVALID_API_KEY);
+        } else if ( $statusCode == 404 ) {
+            throw new KlaviyoResourceNotFoundException(self::ERROR_RESOURCE_DOES_NOT_EXIST);
+        } else if ( $statusCode == 429 ) {
+            throw new KlaviyoRateLimitException( $this->decodeJsonResponse( $response ) );
+        } else if ( $statusCode != 200 ) {
+            throw new KlaviyoException( sprintf( self::ERROR_NON_200_STATUS, $statusCode ) );
+        }
+
+        if ( $isPublic ) {
+            return '1' == $response->getBody();
+        }
+
+        return $this->decodeJsonResponse( $response );
+    }
+
+    /**
+     * Handle authentication by updating $options passed into request method
+     * based on type of API request.
+     *
+     * @param array $options Options configuration for Request Interface
+     * @param bool $isPublic Request type - public
+     * @param bool $isV1 Request API version - V1
+     * @param bool $isV2 Request API version - V2
+     */
+    private function prepareAuthentication ( &$options, $isPublic, $isV1, $isV2 )
```

Comment:
could we just pass in a version arg and then have a helper function like:

if isPublic:
    $this->publicAuth
else if v1:
    $this->v1Auth
else if v2:
    $this->v2Auth


this should also just return something, i don't like the global scoping, let's explicitly return the thing